polaris.shopify.com/src/utils/search.ts

```diff
@@ -149,42 +141,14 @@ export function search(query: string): GroupedSearchResults {
       score: result.score || 0,
     }));
 
-    groupedResults["Foundations"].results = scoredResults
-      .filter((result) => result.category === "Foundations")
-      .map((result) => ({
-        ...result,
-        score: result.score || 0,
-      }))
-      .slice(0, MAX_RESULTS["Foundations"]) as FoundationsSearchResult[];
-
-    groupedResults["Components"].results = scoredResults
-      .filter((result) => result.category === "Components")
-      .map((result) => ({
-        ...result,
-        score: result.score || 0,
-      }))
-      .slice(0, MAX_RESULTS["Components"]) as ComponentsSearchResult[];
-
-    groupedResults["Tokens"].results = scoredResults
-      .filter((result) => result.category === "Tokens")
-      .map((result) => ({
-        ...result,
-        score: result.score || 0,
-      }))
-      .slice(0, MAX_RESULTS["Tokens"]) as TokensSearchResult[];
-
-    groupedResults["Icons"].results = scoredResults
-      .filter((result) => result.category === "Icons")
-      .map((result) => ({
-        ...result,
-        score: result.score || 0,
-      }))
-      .slice(0, MAX_RESULTS["Icons"]) as IconsSearchResult[];
-
-    Object.keys(groupedResults).forEach((category) => {
-      const typedCategory = category as SearchResultCategory;
-      groupedResults[typedCategory].maxScore =
-        groupedResults[typedCategory].results[0]?.score || 0;
+    scoredResults.forEach((result) => {
+      const categoryResults = groupedResults[result.category].results;
+      if (categoryResults.length < MAX_RESULTS[result.category]) {
+        if (!result) return;
+        // @ts-expect-error
```

Comment:
I would like to clean this file up. It's frustrating that we need to duplicate the same code multiple times because of the types needed by `as IconsSearchResult[]` `as TokensSearchResult[]`...

Any thoughts? cc @aaronccasanova

Reply:
I was able to clean this code up by simplifying the search result types.

The search results are almost identical so they can use the same root type. The main difference is the `meta` object which contains different things depending on what type of search result it is.

New code: https://github.com/Shopify/polaris/pull/6502/files#diff-3e0197d18c6157be3d623c9dd24dc76a4b1b2cc8e300f0f99cbfa25bc85641b4R172

Simplified types: https://github.com/Shopify/polaris/pull/6502/files#diff-45904998eafd7e76b04406ff0b277701c8732c59ab9bd8ab93b8518118769497R48